sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,655 @@
+----
+|
+Allows assignment from a pointer to type `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator=(annotated_ptr const&) noexcept;
+----
+|
+Allows assignment from an `annotated_ptr` with the same parameterization.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator++() noexcept;
+----
+|
+Prefix increment operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator++() noexcept;
+----
+|
+Postfix increment operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator--() noexcept;
+----
+|
+Prefix decrement operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator--() noexcept;
+----
+|
+Postfix decrement operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr bool has_property();
+----
+| 
+Returns true if the property list contains the property with property key class
+`propertyT`. Returns false if it does not.
+
+Available only when `propertyT` is a property key class.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr auto get_property();
+----
+| 
+Returns a copy of the property value contained in the property list
+`PropertyListT`. Must produce a compile error if `PropertyListT` does not
+contain a property with the `propertyT` key.
+
+Available only if `propertyT` is the property key class of a compile-time
+constant property.
+|===
+
+=== Add new reference wrapper class `annotated_ref` to enable `annotated_ptr`
+
+The purpose of the `annotated_ref` class template is to provide reference
+wrapper semantics. It enables the implementation to preserve the properties on
+loads from and stores to the pointers.
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ref {
+  public:
+    annotated_ref(T *);
+    operator T() noexcept;
+    operator const T() const noexcept;
+    void operator=(const T &);
+  private:
+    T *Ptr;
+  };
+} // namespace sycl::ext::oneapi
+```
+
+
+Member Functions are described in the table below
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ref(T * InputPtr);
+----
+|
+Constructs an `annotated_ref` object. Does not allocate new storage. The
+underlying pointer is initialized to `InputPtr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+operator T() noexcept;
+operator const T() const noexcept;
+----
+|
+Reads the object of type `T` that is referenced by this wrapper, applying the
+annotations when the object is loaded from memory.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+void operator=(const T &);
+----
+|
+Writes an object of type `T` to the location referenced by this wrapper,
+applying the annotations when the object is stored to memory.
+
+|===
+
+=== Properties
+
+Below is a list of compile-time constant properties supported with
```

Comment:
If we want to allow properties with runtime values in the future then this should also supply a non-static `get_property` overload. It could be added with the corresponding properties, but it would be cleaner to have it ready.

Reply:
I don't think annotated_ptr should support runtime properties.  Then those properties would have to be kept around in an additional member variable in the struct.  We want the class to be a wrapper around only a pointer.